### ISSUE TITLE
Won't stay running in the background with Android 6/7

### ISSUE THREAD

**Issue by luke-jr** [tananaev/traccar-client-android#240]
Traccar seems to work great on Android 4.1 most of the time, but with Android 6 and 7 it stops working when it's in the background. Android settings are configured to allow it background data access, so that's not the problem. :/

**Comment by tananaev** [comment]
I'm guessing that your phone is configured to use Doze mode? You have to add Traccar Client into exceptions.

**Comment by luke-jr** [comment]
I don't believe so, but I have never heard of this mode. How do I check?

**Comment by luke-jr** [comment]
Also, I reproduced the problem without it even turning the display off. One phone had 2 weeks worth of GPS updates to upload, but as soon as I hit "home", it stopped sending them...

**Comment by tananaev** [comment]
That might be some vendor-specific issue as well.

Doze mode is enabled by default for all apps. Go into settings -> battery -> battery optimization (in the menu) -> check if app is in the list. If not, it means that Android can kill it to save battery life.

**Comment by luke-jr** [comment]
Bingo, I think. At least two of the affected phones were "optimising" it away. Will hopefully find conclusive results over the next week... Thanks.

I don't think there are any vendor-specific issues possible. The three phones experiencing this issue are a Nexus 5X from Google's Project Fi running stock Android 7, and two Moto E 2015 with Cyanogenmod 13 (Android 6).

**Comment by tananaev** [comment]
If you are using stock or close to stock Android version, there should be no vendor specific issues. Let me know how it goes.

**Comment by luke-jr** [comment]
Nope, doesn't seem to have helped at all. :(

**Comment by tananaev** [comment]
I guess the only way to know what happened is to get device logs.

**Comment by luke-jr** [comment]
What's the best way to get them?

**Comment by tananaev** [comment]
You need to get Android SDK and use ADB USB connection to get logs. Android Studio might be easier than command line tools.

**Comment by luke-jr** [comment]
Android SDK/Studio are non-free, so I'm not really interested in using them, but I have ADB...

**Comment by tananaev** [comment]
What do you mean non-free? They are free. ADB is a part of the SDK, so if you have ADB, it likely means that you have SDK as well.

**Comment by luke-jr** [comment]
I mean they have restrictive EULAs that do not meet the definition of [free software](https://www.gnu.org/philosophy/free-sw.en.html).

Gentoo has a dev-util/android-tools package with adb, fastboot, and mkbootimg on their own.

**Comment by tananaev** [comment]
I see. Then yes, you should be able to use `adb` from there to get logs.

**Comment by luke-jr** [comment]
```
130|root@surnia:/ # logcat -d |grep -i tracc                                                                
11-25 02:33:16.622 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 357 lines
11-25 02:33:38.749 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 238 lines
11-25 02:33:50.628 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 365 lines
11-25 02:34:12.380 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 497 lines
11-25 02:34:37.080 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 337 lines
11-25 02:34:53.326 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 226 lines
11-25 02:35:04.649 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 162 lines
11-25 02:35:15.284 25965 25965 I chatty  : uid=1010083(org.traccar.client) expire 85 lines
root@surnia:/ # date
Fri Nov 25 02:35:58 EST 2016
```

Is there something more useful to look for?

**Comment by tananaev** [comment]
I'm not sure about keywords. You should probably get a fragment of the logs from the time around the time when app dies.

**Comment by luke-jr** [comment]
Having trouble reproducing it as I watch the logs :(

**Comment by luke-jr** [comment]
With the 30 second interval set, it was reliably reporting all weekend (at home). When my wife took the phone with her today, it stopped reporting pretty much immediately when she left, and sent the reports immediately when she returned... Unfortunately, outside the home, I have no way to get the logcat... :/

**Comment by tananaev** [comment]
I'm going to shamelessly put link my other app here:

https://play.google.com/store/apps/details?id=com.tananaev.logcat

It might help with getting logs when device is not connected. App is completely open source as well and uses local adb connection on the phone.

**Comment by luke-jr** [comment]
Turns out the log is pretty long... and includes the entire duration my wife was gone :)

https://drive.google.com/file/d/0B4vQXt6BTVAWbFo2Nm1Sam5XRkU/view?usp=sharing

Not sure if there's sensitive data in this, so it's set to require approval of access. Just request access and I will approve it.

**Comment by tananaev** [comment]
Requested access from my gmail account.

**Comment by tananaev** [comment]
There is nothing from the app itself in the logs, but there are some "am_pss" logs with traccar client id in them.

Here is the info about "am_pss":
```
# Report collection of memory used by a process
30047 am_pss (Pid|1|5),(UID|1|5),(Process Name|3),(Pss|2|2),(Uss|2|2),(SwapPss|2|2)
```

I'm assuming those logs mean that client process is running and wasn't killed, but for some reason it doesn't do anything. It must be one of the following things:

1. Device doesn't have any location info and therefore it doesn't update anything
2. Service is on "OFF" state

I guess you've checked the second one already, so it must be something wrong with location on the device.

**Comment by luke-jr** [comment]
Once the device got home, it uploaded updates of its position throughout the entire trip. I'm pretty sure that should rule out any problems with location info. The service is definitely not off either... :/

In the past, I noticed sometimes restarting the service made it start working when it formerly had not been. Is it possible that it's trying to reuse a TCP connection or something?

**Comment by tananaev** [comment]
Usually it means that there was no internet connection.

Reusing same TCP connection shouldn't be a problem. Client uses HTTP protocol, so it fails quickly if it doesn't get an expected response from the server.

**Comment by luke-jr** [comment]
There's definitely an internet connection. Sometimes (but not always) "Where's My Droid" does function while Traccar isn't.

If a TCP connection is trying to use an old WiFi source address, I wouldn't expect its packets to be accepted over the CDMA network, so it would need to wait for a timeout... Also, my home network is IPv6, and it appears FreedomPop/Sprint's CDMA connection does not - so even if it would have forwarded or rewritten the source address, it is absolutely impossible for an IPv6 packet to go out over an IPv4 link.

**Comment by tananaev** [comment]
Are you sure the logs that you provided include everything and it's not just the system logs? How did you get them?

**Comment by luke-jr** [comment]
`logcat -b all -C -v printable`

**Comment by tananaev** [comment]
With `adb` via USB cable?

**Comment by luke-jr** [comment]
SSH'd in, but it should be the same shell...

**Comment by tananaev** [comment]
Directly from the phone? It's NOT the same.

**Comment by luke-jr** [comment]
No, I ssh'd from my desktop PC to the phone... How should I get it then, exactly?

Here's from `adb logcat` with the phone's ADB server running as root (I couldn't figure out how to get it to restart adb as non-root...): https://drive.google.com/file/d/0B4vQXt6BTVAWbFZrcmItQ2hpSUE/view?usp=sharing

**Comment by tananaev** [comment]
You should use `adb` client from your desktop, not though ssh.

**Comment by luke-jr** [comment]
Yes, that last one was adb from my desktop.

**Comment by tananaev** [comment]
Unfortunately it doesn't have any information about the app at all. I'm thinking now if I should implement my own log file. That would help with investigating any issues.

**Comment by luke-jr** [comment]
Anything new I can try? Can't make any promises yet, but if I can send you one of these phones with US cellular service (to keep), would that help ensure it is well-supported?

**Comment by tananaev** [comment]
Well, you have to get logs, including system and the app logs. I won't be able to help without it.

Having the device would certainly help with debugging, but I can't promise anything.

**Comment by luke-jr** [comment]
I provided all logs I know how to get several months ago. :(

**Comment by tananaev** [comment]
Those logs didn't have anything about the app, which means that those were cut down logs. I need full ones. The easiest way to get full logs is via USB through ADB.

**Comment by tananaev** [comment]
There was absolutely no info about the app. If it's full logs, the app was not running on the phone.

**Comment by luke-jr** [comment]
It's everything from `adb logcat` with a phone running Traccar client as the same user, and saying the service is active. :/

**Comment by tananaev** [comment]
What do you mean by "as the same user"?

**Comment by luke-jr** [comment]
The phone is logged in to the same user that is running Traccar.

**Comment by tananaev** [comment]
I don't think it's relevant if you use ADB through USB. I think ADB daemon always runs as a root anyway.

**Comment by luke-jr** [comment]
FYI I put LineageOS 14.1 (equiv to Android 7.1) on a few of these and it seems to be doing better... but time will tell.

**Comment by luke-jr** [comment]
This seems to mostly be resolved these days. The only remaining issues are:

* Sometimes the phone itself doesn't retain connectivity when changing from wifi to mobile. This is obviously an Android problem, not Traccar.
* When Traccar upgrades itself, the service is stopped.

**Comment by tananaev** [comment]
I don't think there is any way to solve the second issue. I am not aware of any way to start the app automatically after installation or upgrade.

**Comment by luke-jr** [comment]
What about this `android.intent.action.MY_PACKAGE_REPLACED` intent thing?

**Comment by tananaev** [comment]
Looks like you are right. I will implement it.

**Comment by tananaev** [comment]
Implemented in a9d772bfda818e26fcd89b31ff96a8799f9d94aa. I will upload new version to the Google Play soon.
